lib/lambda.js

```diff
@@ -2,11 +2,275 @@
+  return !!(event && event.requestContext && event.requestContext.requestId)
+}
+
+function setSqsData (agent, transaction, event, context, isColdStart) {
+  const record = event && event.Records && event.Records[0]
+  record.eventSourceARN = record.eventSourceARN ? record.eventSourceARN : ''
+
+  const faas = getSharedFaasData(context, isColdStart)
+  faas.trigger.request_id = record.messageId
+  faas.trigger.type = 'pubsub'
+  transaction.setFaas(faas)
+
+  const arnParts = record.eventSourceARN.split(':')
+  const queueName = arnParts[5]
+  const accountId = arnParts[4]
+
+  transaction.name = `RECEIVE ${queueName}`
+  transaction.type = 'messaging'
+
+  const serviceContext = {
+    origin: {
+      name: queueName,
+      id: record.eventSourceARN
+    }
+  }
+  transaction.setServiceContext(serviceContext)
+
+  const cloudContext = {
+    origin: {
+      provider: 'aws',
+      region: record.awsRegion,
+      service: {
+        name: 'sqs'
+      },
+      account: {
+        id: accountId
+      }
+    }
+  }
+  transaction.setCloudContext(cloudContext)
+
+  let age
+  if (record.attributes && record.attributes.SentTimestamp) {
+    age = Date.now() - record.attributes.SentTimestamp
+  }
+  const messageContext = {
+    queue: record.eventSourceARN,
+    age: age,
+    body: undefined,
+    headers: Object.assign(record.attributes ? record.attributes : {})
+  }
+
+  if (agent._conf.captureBody) {
+    messageContext.body = record.body
+  }
+  transaction.setMessageContext(messageContext)
+}
+
+function isSqsEvent (event) {
+  const records = event.Records ? event.Records : []
+  return records.length > 0 && records[0].eventSource === 'aws:sqs'
+}
+
+function setSnsData (agent, transaction, event, context, isColdStart) {
+  const record = event && event.Records && event.Records[0]
+  const message = record && record.Sns
+  const faas = getSharedFaasData(context, isColdStart)
+  faas.trigger.type = 'pubsub'
+  faas.trigger.request_id = record && record.Sns && record.Sns.MessageId
+
+  transaction.setFaas(faas)
+
+  let topicArn = record && record.Sns && record.Sns.TopicArn
+  topicArn = topicArn || ''
+  const arnParts = topicArn.split(':')
+  const topicName = arnParts[5]
+  const accountId = arnParts[4]
+  const region = arnParts[3]
+
+  transaction.name = `RECEIVE ${topicName}`
+  transaction.type = 'messaging'
+
+  const serviceContext = {
+    origin: {
+      name: topicName,
+      id: topicArn,
+      version: record.EventVersion
+    }
+  }
+  transaction.setServiceContext(serviceContext)
+
+  const cloudContext = {
+    origin: {
+      provider: 'aws',
+      region: region,
+      service: {
+        name: 'sns'
+      },
+      account: {
+        id: accountId
+      }
+    }
+  }
+  transaction.setCloudContext(cloudContext)
+
+  let age
+  if (message.Timestamp) {
+    age = Date.now() - (new Date(message.Timestamp).getTime())
+  }
+  const messageContext = {
+    queue: topicArn,
+    age: age,
```

Comment:
Same as for SQS above: Is the apm.git lambda spec wrong? The APM server transaction.json spec says `.queue.name` and `.age.ms`.

Reply:
I'm going to honor the APM Server spec for the time being since that's the code that's deployed to the world at the moment. changed in dcc73aae2961b6a9ba1a6b48ff682ab7bb566bb1